ts/a11y/SpeechMenu.ts

```diff
@@ -0,0 +1,211 @@
+ * @param {string} locale The current locale.
+ * @return The menu of the smart choices.
+ */
+function smartPreferences(previous: string, smart: string, locale: string) {
+  const prefs = Sre.clearspeakPreferences.getLocalePreferences();
+  const loc = prefs[locale];
+  if (!loc) {
+    return [];
+  }
+  const items = [
+    {type: 'label', content: 'Preferences for ' + smart},
+    {type: 'rule'}
+  ];
+  return items.concat(loc[smart].map(function (x) {
+    const [key, value] = x.split('_');
+    return {
+      type: 'radioCompare',
+      content: value,
+      id:
+      'clearspeak-' + Sre.clearspeakPreferences.addPreference(previous, key, value),
+      variable: 'speechRules',
+      comparator: (x: string, y: string) => {
+        if (x === y) {
+          return true;
+        }
+        if (value !== 'Auto') {
+          return false;
+        }
+        let [dom1, pref] = x.split('-');
+        let [dom2] = y.split('-');
+        return dom1 === dom2 && !Sre.clearspeakPreferences.fromPreference(pref)[key];
+      }
+    };
+  }));
+}
+
+/**
+ * Creates dynamic clearspeak menu.
+ * @param {MJContextMenu} menu The context menu.
+ * @param {Submenu} sub The submenu to attach elements to.
+ */
+export function clearspeakMenu(menu: MJContextMenu, sub: Submenu) {
+  let locale = menu.pool.lookup('locale').getValue() as string;
+  const box = csSelectionBox(menu, locale);
+  let items: Object[] = [];
+  let explorer = (menu.mathItem as ExplorerMathItem)?.
+    explorers?.explorers?.speech as SpeechExplorer;
+  if (explorer?.walker) {
+    let semantic = explorer.walker.getFocus()?.getSemanticPrimary();
+    if (semantic) {
+      const previous = Sre.clearspeakPreferences.currentPreference();
+      const smart = Sre.clearspeakPreferences.relevantPreferences(semantic);
+      items = items.concat(basePreferences(previous));
+      items = items.concat(smartPreferences(previous, smart, locale));
+    }
+  }
+  if (box) {
+    items.splice(2, 0, box);
+  }
+  return menu.factory.get('subMenu')(menu.factory, {
+    items: items,
+    id: 'Clearspeak'
+  }, sub);
+}
+MJContextMenu.DynamicSubmenus.set('Clearspeak', clearspeakMenu);
+
+/**
+ * Creates dynamic locale menu.
+ * @param {MJContextMenu} menu The context menu.
+ * @param {Submenu} sub The submenu to attach elements to.
+ */
+export function localeMenu(menu: MJContextMenu, sub: Submenu) {
+  let radios: {type: string, id: string,
+               content: string, variable: string}[] = [];
+  for (let lang of Sre.locales.keys()) {
+    if (lang === 'nemeth') continue;
+    radios.push({type: 'radio', id: lang,
+                 content: Sre.locales.get(lang) || lang, variable: 'locale'});
+  }
+  radios.sort((x, y) => x.content.localeCompare(y.content, 'en'));
+  return menu.factory.get('subMenu')(menu.factory, {
+    items: radios, id: 'Language'}, sub);
+}
+MJContextMenu.DynamicSubmenus.set('A11yLanguage', localeMenu);
```

Comment:
The `DynamicSubmenus` are run whenever the menu is posted.  Does the locale menu have to be regenerated every time?  Or can it just be created dynamically when the Menu object is created and then be used from there.

My understanding is that the dynamic menus are for things that change from equation to equation (like the annotations), but global preferences and such should be able to put in the global menu once and continued to be used for all expressions.  But I may be missing something.

Reply:
This works here. We now avoid recomputing the menu.